# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/workers/WorkerRunnable.cpp
# Commit: b7d4fb0fcc54
# Full Hash: b7d4fb0fcc54c929650660f1c61b63bb4694766f
# Author: Eden Chuang <echuang@mozilla.com>
# Date: 2024-06-25 09:36:02
# Regressor Bug: 1894231
# File Overlap Count: 1
# Description:
#   Bug 1894231 - P11 Remove WorkerThreadRunnable::mWorkerPrivateForPreStartCleaning. r=dom-worker-reviewers,asuth
#   
#   This patch introduces a boolean WorkerThreadRunnable::mCleanPreStartDispatching to indicate if the WorkerThreadRunnable needs to skip its execution in its Run() because of WorkerPrivate::RunLoopNeverRan().
#   
#   If any WorkerPrivate initialization fails in [[ https://searchfox.org/mozilla-central/rev/a26e972e97fbec860400d80df625193f4c88d64e/dom/workers/RuntimeService.cpp#2090-2120 | WorkerThreadPrimaryRunnable::Run() ]], corresponding WorkerJSContext might not be created or created then destroyed before entering [[ https://searchfox.org/mozilla-central/rev/a26e972e97fbec860400d80df625193f4c88d64e/dom/workers/RuntimeService.cpp#2085 | WorkerPrivate::RunLoopNeverRan() ]]. This invalidates the WorkerJSContext, and GetCurrentThreadWorkerPrivate() will not work correctly. In this case, any WorkerThreadRunnable should not run in this invalid environment, so we can just skip the execution by returning NS_OK in its Run().
# ==============================================================================

diff -r bb302697f033 -r b7d4fb0fcc54 dom/workers/WorkerRunnable.cpp
--- a/dom/workers/WorkerRunnable.cpp	Mon Jun 24 22:20:54 2024 +0000
+++ b/dom/workers/WorkerRunnable.cpp	Mon Jun 24 22:20:54 2024 +0000
@@ -366,11 +366,15 @@
 NS_IMETHODIMP
 WorkerThreadRunnable::Run() {
   LOG(("WorkerThreadRunnable::Run [%p]", this));
+
+  // The Worker initialization fails, there is no valid WorkerPrivate and
+  // WorkerJSContext to run this WorkerThreadRunnable.
+  if (mCleanPreStartDispatching) {
+    LOG(("Clean the pre-start dispatched WorkerThreadRunnable [%p]", this));
+    return NS_OK;
+  }
+
   WorkerPrivate* workerPrivate = GetCurrentThreadWorkerPrivate();
-  if (!workerPrivate && mWorkerPrivateForPreStartCleaning) {
-    workerPrivate = mWorkerPrivateForPreStartCleaning;
-    mWorkerPrivateForPreStartCleaning = nullptr;
-  }
   MOZ_ASSERT_DEBUG_OR_FUZZING(workerPrivate);
 #ifdef DEBUG
   workerPrivate->AssertIsOnWorkerThread();